### ISSUE TITLE
-ss didn't work perfectly

### ISSUE THREAD

**Issue by MatthiasDING** [PokemonGoMap/PokemonGo-Map#889]
In my case, Hex grid scan got almost 2x more pokemons than spawn scan (with -ss option).  Those two scans were applied with same location, same scan limit,  same number of account.

Does anyone have the same problem?

**Comment by rrbailey89** [comment]
Did you let the beehive scan run for an hour or two prior to use -ss ?

**Comment by MatthiasDING** [comment]
@rrbailey89 Sure, I have run Hex grid scan about 4 hours and got 567 spawn points for using -ss

**Comment by blump** [comment]
I have been testing all afternoon with the -ss option without ss option ( I have done in the hive without scanning option ss for an hour and I stopped when it c is blocked at 10634 points). And I can say that I scan a lot less pokemon with the ss option and some poke 5 or 7 mins ...

**Comment by blump** [comment]
a difference of 40 % of Pokemon . There without ss I am about 1630 to 1000 Pokemon with the ss option ...

**Comment by mrfantomex** [comment]
The -ss option is not working as advertised. Try using --dump-spawns even if you don't need to dump them and it will pick up missing spawn points.

**Comment by blump** [comment]
@mrfantomex thx 
I'll test this afternoon

**Comment by carze** [comment]
The problem with --dump-spawnpoints is that it will dump the contents of your whole database and won't just scan the points that you may pass it or the location you may provide it.

**Comment by BaIthamel** [comment]
You're not proving you found 2x as many pokemon with hex over -ss do that then I'll look into it.

**Comment by blump** [comment]
@carze 
I have a file .json per worker which contains the spawn points after 1 hour and 30 minutes without scanning -ss .

So I would have to inject on the fly new spawn points found with the -ss option in .json files ...

**Comment by carze** [comment]
@blump When I tried -ss with spawn points split between workers through different json files it went through all my spawn points in my current database and dumped them to a new file which was used as the list of spawns to search with.

I suppose this might work if you have a clean database and start up all your workers at the same time unless I am missing something on how this is supposed to work.

**Comment by blump** [comment]
@carze How you use the new file ? Sorry I do not understand . Thank you for your reply .

**Comment by carze** [comment]
@blump I have five workers each with their own unique spawns.json file that cover an area of my city. These spawns were generated using a beehive scan of the area I am covering with my five workers so those coordinates are in my database and the spawns.json files were created from the database.

When I run each worker with its own unique spawns.json file in conjunction with the --dump-spawnpoints flag each of the workers spawns.json files is overwritten with the collection of all spawns from the database (which is what the flag does). Now each of the five workers searches all of the spawn points instead of the regions that I had originally wanted them too.

**Comment by blump** [comment]
@carze thx

**Comment by drewreece** [comment]
@carze, it sounds like you need to specify a unique spawns.json to each command. Are you setting a different file in the arguments?
`runserver.py -ss spawns-1.json -l loc1 -st x
runserver.py -ss spawns-2.json -l loc2 -st x …
`
You can use the -st stepsize to grab spawns in a beehive area via --dump-spawnpoints or you can consider using an external tool to grab rectangles of spawnpoints…
https://gist.github.com/Xcelled/c5b397dfe2c32e61e7aa06479b32bc34
That accepts co-ordinates as input & spews out a the json which can be redirected to whatever file you like. It was adapted into the -ss feature.

**Comment by mrfantomex** [comment]
> You can use the -st stepsize to grab spawns in a beehive area via --dump-spawnpoints

What he's saying is you actually can't do this. It doesn't work. It ignores the -st and just dumps the whole database. This is consistent with my experience.

**Comment by blump** [comment]
@mrfantomex for me -st works fine  to grab spawns

**Comment by drewreece** [comment]
Sorry, this is not consistent with my experience :)

I have the latest version from develop (77a4490)
If I specify the step size, a location and a spawn file I get a limited amount of spawns as expected.

Please post the command you are running to be sure that your args are not part of the issue.

**Comment by blump** [comment]
For me : Templates Ansible :
Create spawn.json by worker  : python /var/www/runserver.py -ss /home/{{user}}/spawn_json/{{auto[0]}}.json -u {{user_prefix}}{{count}} -p {{password}} -l "{{auto[0]}}, {{auto[1]}}" -ns -st {{st}} --dump-spawnpoints

Then, launch worker with : 
python /var/www/runserver.py -ss /home/{{user}}/spawn_json/{{auto[0]}}.json -u {{user_prefix}}{{count}} -p {{password}} -l "{{auto[0]}}, {{auto[1]}}" -ns

**Comment by carze** [comment]
@drewreece I tried using -st, -l and -ss spawns.json all together earlier today and it ended up overwriting the spawns.json (which contained a subset of all my spawns) with all the spawns from my database and then proceeded to check all of those spawn points.

I am using develop commit 36659fc and saw this behaviour. Is the example runserver.py you posted to me earlier what you run when it all works correctly for you?

**Comment by MatthiasDING** [comment]
@drewreece Do you means that for a large area, we should divide this area into several small regions? and for each region, we could get a spawn file (specify st, location), then send it respectively to a server instance.
But in my case, I run a server instance using -st 11, I think it is not so large.  Anyway, how many accounts did you use for a single server instance? @drewreece

**Comment by drewreece** [comment]
@carze, I don't understand why it would dump the entire db, sorry. I used a -st of 2 & got 30 spawns, I also did -st 8 & got 368 spawns. The .jsons looked correct (I didn't count the spawns but the content differed significantly).
 
I would consider adding more log.info() statements to runserver.py…
https://github.com/PokemonGoMap/PokemonGo-Map/blob/ddd10cf8d6f8737e30b1fe73b7157f609f6eb0a8/runserver.py#L187
To see what data it is passing to Pokemon.get_spawnpoints_in_hex()
If you haven't done so already be sure you have installed the latest requirements & done a rebuild, sometimes older files can cause weird issues ask if you are unsure, I think is is just `pip install -r requirements.txt --upgrade` and a `grunt build` 

I'd also clone again to a new directory & duplicate over your config & DB as a test if you can stand the extra effort.
Use Xcelled's spawn-exporter.py if you need a workaround.

@MatthiasDING, I don't scan large areas, I believe it will get you banned & puts too much load on Niantic :)
I do think it is wise to create a spawns.json for a small area & then break up the batches when handling large areas with spawn scanning. In theory it will stop accounts teleporting large distances. There has been discussion on 'speed limiting' workers but I don't think it has been added yet. FYI I used 2 accounts on a step size of 8, delay 10, it works better with spawn scanning, in this area beehive is a little limited (spawns in most cells, but spread over time).

**Comment by blump** [comment]
After a test after noon , the -ss option that works less well without the option.
My config :
- 200 workers (with 1 pokemon caught ), 3 Vps
- 6 gold st st 4
- SD 7.2 ( works fine )
- mysql
- Restarting workers randomly between 26 and 29 mins with a sleep between 110 and 123 seconds

**Comment by blump** [comment]
the problem may be the ram memory. Without -ss 60 workers about I to 4 GB with ss I get to 8GB, so there just saturation server can be the problem ... I'll test on a server 64GB to be safe

**Comment by blump** [comment]
Finally after many test , I return to what I said. The -ss option works very well and much better than a scan without the option. But the -ss option uses much more memory ram . I had to halve the number of worker . Before I had a config -st 6 and -lp 3 , now I -st 9 and - lp 2. And it works wonders for me .

**Comment by MatthiasDING** [comment]
@blump why you have option -lp ？

**Comment by blump** [comment]
@MatthiasDING 
it is for the generation of ruches automatically with "Tools/Hex-Beehive-Generator/location_generator.py"

**Comment by BaIthamel** [comment]
you still haven't shown an issue. if you can't show me an actual issue before i check again tomorrow i'll close this.

**Comment by MatthiasDING** [comment]
@BaIthamel  Ok, to verify this issue, I started two server instances at the same time(one with -ss) respectively on two mysql databases. 

For the database used with -ss option, I have already run the Hex grid scan about two hours,  and another database was initially empty.

Those two servers were started with same configrations (location, step limit, scan delay, number of account) except the -ss option.

Here are the printed status from two server instances:
without -ss
> Queue: 103 items
> Overseer: Processing search queue
> Worker 000 - Success: 38, Failed: 0, No Items: 85 - Waiting 10 seconds for scan delay
> Worker 001 - Success: 37, Failed: 0, No Items: 85 - Waiting 10 seconds for scan delay
> Worker 002 - Success: 29, Failed: 0, No Items: 93 - Waiting 10 seconds for scan delay
> Worker 003 - Success: 28, Failed: 0, No Items: 94 - Waiting 10 seconds for scan delay
> Worker 004 - Success: 32, Failed: 0, No Items: 90 - Waiting 10 seconds for scan delay
> Worker 005 - Success: 41, Failed: 0, No Items: 81 - Waiting 10 seconds for scan delay
> Worker 006 - Success: 37, Failed: 0, No Items: 85 - Waiting 10 seconds for scan delay
> Worker 007 - Success: 40, Failed: 0, No Items: 82 - Waiting 10 seconds for scan delay
> Worker 008 - Success: 38, Failed: 0, No Items: 84 - Waiting 10 seconds for scan delay
> Worker 009 - Success: 33, Failed: 0, No Items: 89 - Waiting 10 seconds for scan delay

with -ss

> Queue: 0 items
> Overseer: Waiting for spawnpoints 327 of 457 to spawn at 2544
> Worker 000 - Success: 13, Failed: 0, No Items: 3, Skipped: 0 - Waiting for item from queue
> Worker 001 - Success: 13, Failed: 0, No Items: 3, Skipped: 0 - Waiting for item from queue
> Worker 002 - Success: 9, Failed: 0, No Items: 7, Skipped: 0 - Waiting for item from queue
> Worker 003 - Success: 12, Failed: 0, No Items: 3, Skipped: 0 - Waiting for item from queue
> Worker 004 - Success: 6, Failed: 0, No Items: 10, Skipped: 0 - Waiting 10 seconds for scan delay
> Worker 005 - Success: 8, Failed: 0, No Items: 7, Skipped: 0 - Waiting for item from queue
> Worker 006 - Success: 9, Failed: 0, No Items: 7, Skipped: 0 - Waiting for item from queue
> Worker 007 - Success: 10, Failed: 0, No Items: 5, Skipped: 0 - Waiting for item from queue
> Worker 008 - Success: 12, Failed: 0, No Items: 3, Skipped: 0 - Waiting for item from queue
> Worker 009 - Success: 8, Failed: 0, No Items: 7, Skipped: 0 - Waiting for item from queue

The numbers of success were definitely different.

Here are two maps' screenshot:
without -ss
![without_ss](https://cloud.githubusercontent.com/assets/6976164/17840922/6ed9f442-6812-11e6-8067-a1e3b0a9456e.png)

with -ss:
![with_ss](https://cloud.githubusercontent.com/assets/6976164/17840924/76b9f40a-6812-11e6-8817-df2fb19cad62.png)

**Comment by mrfantomex** [comment]
MatthiasDING can you now run the -ss server with --dump-spawnpoints and see if that resolves the issue, making note of the number of spawn points it says early on in the log?

This is the workaround that I'm using. I'm not sure -ss is meant to be used without either --dump-spawnpoints or a .json file input. This might just be a documentation issue.

**Comment by mrfantomex** [comment]
Also, -ss has a bunch of other problems I assume someone is aware of....

1.) Pausing in the UI does not work.

2.) Moving the location in the UI does not work.

**Comment by MatthiasDING** [comment]
@mrfantomex using --dump-spawnpoints didn't resolve this issue. I got the same same number of spawn points as that without --dump-spawnpoints.  In my case, the database only contained scanned data from a small region with option -st 11.

**Comment by MatthiasDING** [comment]
Looking at the printed status with -ss option,  I found that it had many times of "No Items".  Maybe that is useful information.

**Comment by carze** [comment]
@MatthiasDING 'No Items' to me seems like it could be a throttling issue? I see less 'No Item' entries when I do spawn point scanning but it seems to me there are less pokemon scanned as well. 

If I get a chance this afternoon I will try to run a test as well on a small subset of my spawn points compared against traditional scanning.

**Comment by MatthiasDING** [comment]
@carze  I think for Beehive scanning, "No items" is acceptable because it just walk through all possible locations. But for spawn point scanning, why I got many "No items" ? it's maybe the cause.

**Comment by blump** [comment]
Create spawn points this morning for 3 hours.

Why so many empty green circle?
-sd 10.2
<img width="533" alt="capture d ecran 2016-08-22 a 12 30 11" src="https://cloud.githubusercontent.com/assets/10714122/17851844/5be947dc-6864-11e6-9ff1-2546fafb5beb.png">

**Comment by MatthiasDING** [comment]
I think empty green circle means "No item found".  I have no idea how it happened. 
Maybe this code has something wrong with dealing the spawn time.

**Comment by carze** [comment]
I also have several empty scanned circles in my setup but I don't have many 'No Items' entries so I don't think the 'No Items' counter matches up with the number of empty circles. 

I think I also recall seeing this in the original implementations of spawn scanning so it might be normal behavior?

**Comment by rrbailey89** [comment]
@carze @mathiasding @blump are y'all filtering certain Pokemon from showing on the map?

**Comment by carze** [comment]
@rrbailey89 I was in fact filtering Pokemon (oops!) when I show all commons I only have a few empty circles that likely corresponding to my 'No Item' count. 

So everything looks good to me.

**Comment by blump** [comment]
@rrbailey89 No for me

**Comment by MatthiasDING** [comment]
Today, I tried the newest commit  #901, it still did't resolve the issue. 
I have check manually the spawns' appear time. That's correct.  But spawn scan could not find item while Hex scan could find pokemons at that location. 
Really strange things.

**Comment by MatthiasDING** [comment]
Finally, I fixed this issue myself. I didn't get any "Empty (No Items)" in 30 minutes. As a result, spawn scan got same number of pokemon as Hex grid scan.   

I have done following steps:
1. Update to the newest commit after #901.
2. Modify seach.py in line 461, 464, 468, from "**now() < appears + 10**" to "**now() < appears + 120**". (maybe could set a smaller time interval than 120. but in my case, I have tried 60 and still have "No items")
3. set scan delay (-sd) much bigger. I changed from 10 to 15. 

Now, -ss works perfectly to me.
@blump  I wish my information could help you. 
@BaIthamel @chuyskywalker Is there a flexible way to set that time interval in step 2 ? Could you make a new update? Thank you very much.

**Comment by blump** [comment]
@MatthiasDING thx i weel see to it

**Comment by nikos90** [comment]
@MatthiasDING that worked for me too!
Thanks.

**Comment by chuyskywalker** [comment]
`now() < appears + 120` Might be a good improvement. The `10` was me being worried that our disappear times wouldn't be quite right, so I wanted a little buffer. Scanning after two minutes isn't too bad.

**Comment by MatthiasDING** [comment]
@chuyskywalker Now, I reduce its value from 120 to 90. `now() < appears + 90`also works well.   
but `now() < appears + 60` worked badly for me.

**Comment by lota7** [comment]
I changed value to  +120 on all 3 lines and I still get 0/0/0 on first upsert.

![image](https://cloud.githubusercontent.com/assets/20760362/17883355/5896801e-6912-11e6-9a75-c50d35dcdce7.png)

**Comment by MatthiasDING** [comment]
@lota7  Maybe you should run server 10~20 minutes to see the percentage of Empty .  If it is too high, you could try different time intervals.

**Comment by lota7** [comment]
@MatthiasDING I have now let it now 1 hour now, and I see 2 empty per worker now.
It occurrs on first login/upsert, and after re-auth upsert.
#901 was supposed to fix this but it didnt, so i figured this would help, but it did not. 
Is nobody else getting Empty upsets anymore after #901 and the fix in this issue#?

**Comment by MatthiasDING** [comment]
@lota7 I guess your problem now is not related to scan process but related to login delay. 
For me, I didn't got any empty on login after #901. 

However, I think 2 empty per hour is acceptable ^^.

**Comment by blump** [comment]
@MatthiasDING 
I tested and I prefer your config also thank you

**Comment by lota7** [comment]
@MatthiasDING 
Yeah it's most likely login delay related. 
Do you know where this setting would be?

**Comment by MatthiasDING** [comment]
@lota7 sorry, I have no idea about your issue. Maybe you could open a new issue for discuss.

**Comment by MatthiasDING** [comment]
**Update:**  I found actually this is an issue of time synchronization.

Previously, there was a time error between my server and Pokemon server.  After I run time synchronization using NTP servers, -ss works well even with original implement **now() < appears + 10**.


@chuyskywalker @blump @nikos90

**Comment by nikos90** [comment]
@MatthiasDING 
I confirm. Had the same issue again and just synced the clock with NTP servers.
Running perfectly now!!
